src/main.rs

```diff
@@ -0,0 +1,449 @@
+            let res = extrinsic.dispatch(RuntimeOrigin::signed(origin));
+
+            #[cfg(not(feature = "fuzzing"))]
+            println!("    result:     {res:?}");
+
+            elapsed += now.elapsed();
+        }
+
+        finalize_block(elapsed);
+
+        check_invariants(block, initial_total_issuance);
+    });
+}
+
+fn initialize_block(block: u32, prev_header: Option<&RelayHeader>) {
+    #[cfg(not(feature = "fuzzing"))]
+    println!("\ninitializing block {block}");
+
+    let pre_digest = Digest {
+        logs: vec![DigestItem::PreRuntime(
+            AURA_ENGINE_ID,
+            Slot::from(u64::from(block)).encode(),
+        )],
+    };
+    let parent_header = &Header::new(
+        block,
+        H256::default(),
+        H256::default(),
+        prev_header.map(Header::hash).unwrap_or_default(),
+        pre_digest,
+    );
+    Executive::initialize_block(parent_header);
+
+    // 2. Apply Timestamp
+    Executive::apply_extrinsic(UncheckedExtrinsic::new_bare(RuntimeCall::Timestamp(
+        pallet_timestamp::Call::set {
+            now: u64::from(block) * SLOT_DURATION,
+        },
+    )))
+    .unwrap()
+    .unwrap();
+
+    // 3.  Set up parachain validation data
+    let parachain_validation_data = {
+        let parent_head = HeadData(prev_header.unwrap_or(parent_header).encode());
+
+        let sproof_builder = RelayStateSproofBuilder {
+            para_id: ParachainInfo::get(),
+            included_para_head: Some(parent_head.clone()),
+            current_slot: cumulus_primitives_core::relay_chain::Slot::from(2 * u64::from(block)),
+            ..Default::default()
+        };
+
+        let (relay_storage_root, proof) = sproof_builder.into_state_root_and_proof();
+
+        cumulus_pallet_parachain_system::Call::set_validation_data {
+            data: cumulus_primitives_parachain_inherent::ParachainInherentData {
+                validation_data: cumulus_primitives_core::PersistedValidationData {
+                    parent_head: Default::default(),
+                    relay_parent_number: block,
+                    relay_parent_storage_root: relay_storage_root,
+                    max_pov_size: Default::default(),
+                },
+                relay_chain_state: proof,
+                downward_messages: Default::default(),
+                horizontal_messages: Default::default(),
+            },
+        }
+    };
+
+    Executive::apply_extrinsic(UncheckedExtrinsic::new_bare(RuntimeCall::ParachainSystem(
+        parachain_validation_data,
+    )))
+    .unwrap()
+    .unwrap();
+}
+
+fn finalize_block(elapsed: Duration) -> RelayHeader {
+    #[cfg(not(feature = "fuzzing"))]
+    println!("\n  time spent: {elapsed:?}");
+
+    assert!(elapsed.as_secs() <= 2, "block execution took too much time");
+
+    #[cfg(not(feature = "fuzzing"))]
+    println!("finalizing block");
+    Executive::finalize_block()
+}
+
+fn check_invariants(_block: u32, _initial_total_issuance: Balance) {
```

Comment:
Are there more pallet invariants we can reuse off-the-shelf?

I'm just guessing that various different pallets already have such checks implemented.
Probably in the form of try-state.

Reply:
Using those 2 Hooks:
```rust
    AllPalletsWithSystem::integrity_test();
    AllPalletsWithSystem::try_state(block, TryStateSelect::All).unwrap();
```

I don't think there is more